### ISSUE TITLE
scipy.signal.convolve produces inconsistent and backwards incompatible behavior with method parameter set to fft vs direct

### ISSUE THREAD

**Issue by mattfishburn** [scipy/scipy#7193]
Scipy's convolve method now returns different typed output depending on the value given to the method parameter. 

```
import scipy.signal
import numpy as np
a = np.array([True, ] * 1000000)
b = np.ones((50, ))
bool_output = scipy.signal.convolve(a, b, method='fft')
float_output = scipy.signal.convolve(a, b, method='direct')

bool_output.dtype # evals to dtype('bool')
float_output.dtype # evals to dtype('float64')
```

Note that on my system, scipy.signal.convolve in scipy version 0.18.1 produces a float64 array.

EDIT: Code formatting and arg switches, apologies.

**Comment by Eric89GXL** [comment]
The order of arguments is different, does that matter? It shouldn't, but
just want to more clearly isolate the problem.

**Comment by mattfishburn** [comment]
Hi,

I had tested it both ways, the order didn't matter.

I had originally copy/pasted the wrong code segment into the issue, but edited the snippet a few minutes after submission which fixed the argument order.  Are you looking at the previous edit version?  Apologies for that.

Cheers,
Matt

**Comment by Eric89GXL** [comment]
Yeah I was reading the GitHub email -- the updated example looks good, thanks.

@stsievert do you have time to look into this? I think this comes from #5608.

**Comment by stsievert** [comment]
@Eric89GXL the line to blame is `return out.astype(volume.dtype)`. I think the easiest solution would be to change it to `return out.astype(return_types[first_input.dtype][second_input.dtype])` where `return_types` is a predefined dictionary of in/out types found using 0.18 (with checks if types not in `return_types` and keeping the dictionary small).

We may have introduced similar dtype bugs in when moving to `np.convolve` for 1D inputs in d2265e99c3065dfb04ad45ab4b421ec6657987d8. Looking through [test_signaltools.py] I don't see any tests asserting dtype equality (besides the ones I wrote in #5608). How do we want to handle this?

[test_signaltools.py]:https://github.com/scipy/scipy/blob/2f36b176badcc160739419746790e6f1ec7367eb/scipy/signal/tests/test_signaltools.py

**Comment by stsievert** [comment]
I've written up a quick test script to guess at 0.18 convolve dtype for 1D, but some cases could lead to overflow. For example, `convolve(complex128, complex128) => complex128` in 0.18, not complex256 as we'd expect.

**Comment by e-q** [comment]
Why should `convolve(complex128, complex128) => complex256` if the product of two `complex128` numbers is `complex128`? I imagine we could use `np.result_type(x, y)` to infer the correct output type, as we do in the spectral functions.

**Comment by e-q** [comment]
Closed with 760a369
